src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs

```diff
@@ -279,6 +301,48 @@ private static void OnRequestSendingRequest(WinHttpRequestState state)
                 var serverCertificate = new X509Certificate2(certHandle);
                 Interop.Crypt32.CertFreeCertificateContext(certHandle);
 
+                IPAddress? ipAddress = null;
+                unsafe
+                {
+                    Interop.WinHttp.WINHTTP_CONNECTION_INFO connectionInfo;
+                    Interop.WinHttp.WINHTTP_CONNECTION_INFO* pConnectionInfo = &connectionInfo;
+                    uint infoSize = (uint)sizeof(Interop.WinHttp.WINHTTP_CONNECTION_INFO);
+                    if (NetEventSource.Log.IsEnabled()) NetEventSource.Info(state, $"sizeof(WINHTTP_CONNECTION_INFO)={infoSize}");
+                    if (Interop.WinHttp.WinHttpQueryOption(
+                        state.RequestHandle,
+                        Interop.WinHttp.WINHTTP_OPTION_CONNECTION_INFO,
+                        (IntPtr)pConnectionInfo,
+                        ref infoSize))
+                    {
+                        ReadOnlySpan<byte> RemoteAddressSpan = new ReadOnlySpan<byte>(connectionInfo.RemoteAddress, 128);
+                        AddressFamily addressFamily = (AddressFamily)BitConverter.ToInt16(RemoteAddressSpan.ToArray(), 0);
+                        ipAddress = addressFamily switch
+                        {
+                            AddressFamily.InterNetwork => new IPAddress(BinaryPrimitives.ReadUInt32LittleEndian(RemoteAddressSpan.Slice(4))),
+                            AddressFamily.InterNetworkV6 => new IPAddress(RemoteAddressSpan.Slice(8, 16).ToArray()),
+                            _ => null
+                        };
+                        if (ipAddress != null && NetEventSource.Log.IsEnabled()) NetEventSource.Info(state, $"ipAddress: {ipAddress}");
+
+                    }
+                    else
+                    {
```

Comment:
On what Windows versions is `WINHTTP_OPTION_CONNECTION_INFO` supported? Can we disable `WinHttpQueryOption` queries forever if it fails once?

Reply:
Or run it against a fake handle to create a feature flag: https://github.com/dotnet/runtime/blob/f1901a00dfcfbcf0a4f8b6aa7459fec30846d1ab/src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpTrailersHelper.cs

Reply:
Might worth to add a comment.